src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -938,41 +940,69 @@ DeferErrorIfUnsupportedSublinkAndReferenceTable(Query *queryTree)
-								 "cannot pushdown the subquery",
-								 "Reference tables are not allowed in FROM "
-								 "clause when the query has subqueries in "
-								 "WHERE clause", NULL);
-		}
-		else if (recurType == RECURRING_TUPLES_FUNCTION)
-		{
-			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-								 "cannot pushdown the subquery",
-								 "Functions are not allowed in FROM "
-								 "clause when the query has subqueries in "
-								 "WHERE clause", NULL);
-		}
-		else if (recurType == RECURRING_TUPLES_RESULT_FUNCTION)
-		{
-			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-								 "cannot pushdown the subquery",
-								 "Complex subqueries and CTEs are not allowed in "
-								 "the FROM clause when the query has subqueries in the "
-								 "WHERE clause", NULL);
-		}
-		else
-		{
-			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-								 "cannot pushdown the subquery",
-								 "Subqueries without FROM are not allowed in FROM "
-								 "clause when the outer query has subqueries in "
-								 "WHERE clause", NULL);
-		}
+		/*
+		 * There is a distributed table somewhere in the FROM clause.
+		 *
+		 * In the typical case this means that the query does not recur,
+		 * but there are two exceptions:
+		 *
+		 * - outer joins such as reference_table LEFT JOIN distributed_table
+		 * - FROM reference_table WHERE .. (SELECT .. FROM distributed_table) ..
+		 *
+		 * However, we check all subqueries and joins separately, so we would
+		 * find such conditions in other calls.
+		 */
+		return NULL;
+	}
+
+
+	/*
+	 * Try to figure out which type of recurring tuples we have to produce a
+	 * relevant error message. If there are several we'll pick the first one.
+	 */
+	IsRecurringRangeTable(queryTree->rtable, &recurType);
+
+	if (recurType == RECURRING_TUPLES_REFERENCE_TABLE)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "cannot pushdown the subquery",
+							 "Reference tables are not allowed in FROM "
+							 "clause when the query has subqueries in "
+							 "WHERE clause", NULL);
+	}
+	else if (recurType == RECURRING_TUPLES_FUNCTION)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "cannot pushdown the subquery",
+							 "Functions are not allowed in FROM "
+							 "clause when the query has subqueries in "
+							 "WHERE clause", NULL);
+	}
+	else if (recurType == RECURRING_TUPLES_RESULT_FUNCTION)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "cannot pushdown the subquery",
+							 "Complex subqueries and CTEs are not allowed in "
+							 "the FROM clause when the query has subqueries in the "
+							 "WHERE clause", NULL);
 	}
+	else if (recurType == RECURRING_TUPLES_EMPTY_JOIN_TREE)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "cannot pushdown the subquery",
+							 "Subqueries without FROM are not allowed in FROM "
+							 "clause when the outer query has subqueries in "
+							 "WHERE clause", NULL);
+	}
+
+	/*
```

Comment:
Minor: I cannot think of such a query. Can you show an example?

Reply:
`SELECT (SELECT 1)`. It has sublinks, but rtable is empty.

Reply:
Thanks, I forgot that target list subqueries are also called `sublink`s

Reply:
could also be `SELECT 1 WHERE (SELECT 1 > 0);`